crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1406,14 +1407,56 @@ impl<'db> TypeInferenceBuilder<'db> {
         let ast::StmtAugAssign {
             range: _,
             target,
-            op: _,
+            op,
             value,
         } = assignment;
-        self.infer_expression(value);
-        self.infer_expression(target);
 
-        // TODO(dhruvmanila): Resolve the target type using the value type and the operator
-        Type::Todo
+        // Resolve the target type, assuming a load context.
+        let target_type = match &**target {
+            Expr::Name(target) => self.infer_name_load(target),
+            Expr::Attribute(target) => self.infer_attribute_load(target),
+            _ => self.infer_expression(target),
+        };
+        let value_type = self.infer_expression(value);
+
+        // TODO(charlie): Add remaining branches for different types of augmented assignments.
+        if let (Operator::Sub, Type::Instance(class)) = (*op, target_type) {
+            let class_member = class.class_member(self.db, "__isub__");
+            let call = class_member.call(self.db, &[value_type]);
+
+            return match call.return_ty_result(self.db, AnyNodeRef::StmtAugAssign(assignment), self)
+            {
+                Ok(t) => t,
+                Err(e) => {
+                    self.add_diagnostic(
+                        assignment.into(),
+                        "unsupported-operator",
+                        format_args!(
+                            "Operator `{op}` is unsupported for type `{}`",
+                            target_type.display(self.db),
+                        ),
+                    );
+                    e.return_ty()
+                }
+            };
+        }
+
+        let left_ty = target_type;
+        let right_ty = value_type;
+
+        self.infer_binary_expression_type(left_ty, right_ty, *op)
+            .unwrap_or_else(|| {
+                self.add_diagnostic(
+                    assignment.into(),
+                    "unsupported-operator",
+                    format_args!(
+                        "Operator `{op}` is unsupported between objects of type `{}` and `{}`",
+                        left_ty.display(self.db),
+                        right_ty.display(self.db)
+                    ),
+                );
+                Type::Unknown
+            })
```

Comment:
The logic here is meant to mimic:
```python
if hasattr(a, "__isub__"):
    _value = a.__isub__(b)
    if _value is not NotImplemented:
        a = _value
    else:
        a = a - b
    del _value
 else:
     a = a - b
```

As per https://snarky.ca/unravelling-augmented-arithmetic-assignment/.

Reply:
However, I don't really know how to handle the `NotImplemented` case, if at all -- i.e., how we can determine whether to use `__isub__` or fall back to `a - b`. Can we even detect that from the type system?

Reply:
The convention for dunder methods is that if calling the dunder would return `NotImplemented`, it's annotated "as if it were an illegal call". So `int.__sub__` is annotated as "not accepting" `float` in typeshed, even though actually calling it with a `float` "works fine" (it just returns `NotImplemented`, allowing the operation to fallback to `float.__rsub__`):

https://github.com/astral-sh/ruff/blob/60a2dc53e7b4407c3c0c14093fd8542879997ca8/crates/red_knot_vendored/vendor/typeshed/stdlib/builtins.pyi#L273

```pycon
>>> (42).__sub__(42.0)
NotImplemented
```

Currently, though, we don't do any checking of argument types passed into functions, so we can't implement the "fallback to the other method" logic fully right now. There's lots of TODO comments in our tests for binary arithmetic about this currently.

Reply:
Some example TODOs for you: https://github.com/astral-sh/ruff/blob/60a2dc53e7b4407c3c0c14093fd8542879997ca8/crates/red_knot_python_semantic/resources/mdtest/binary/instances.md?plain=1#L284-L291

Reply:
The way these dunder methods are typically annotated is that their parameter type should reflect the type of argument for which the method will return a value (not `Notimplemented`), and any other type will return `NotImplemented`. So we can detect this from the type system, but currently red-knot can't, because we don't yet have call signature checking (it's something I want to work on this week.)

I'll comment more details above.

Reply:
Oops, didn't see Alex's review before submitting mine.